src/sdk/pynni/nni/compression/torch/pruning/amc/channel_pruning_env.py

```diff
@@ -0,0 +1,497 @@
+                this_comp += flop * self.strategy_dict[idx][1]
+                # also add buffer here (influenced by ratio)
+                this_comp += buffer_flop
+            else:
+                other_comp += flop * self.strategy_dict[idx][0] * self.strategy_dict[idx][1]
+                # add buffer
+                other_comp += buffer_flop * self.strategy_dict[idx][0]  # only consider input reduction
+
+        self.expected_min_preserve = other_comp + this_comp * action
+        max_preserve_ratio = (self.expected_preserve_computation - other_comp) * 1. / this_comp
+
+        action = np.minimum(action, max_preserve_ratio)
+        action = np.maximum(action, self.strategy_dict[self.prunable_idx[self.cur_ind]][0])  # impossible (should be)
+
+        return action
+
+    def _get_buffer_flops(self, idx):
+        buffer_idx = self.buffer_dict[idx]
+        buffer_flop = sum([self.layer_info_dict[_]['flops'] for _ in buffer_idx])
+        return buffer_flop
+
+    def _cur_flops(self):
+        flops = 0
+        for idx in self.prunable_idx:
+            c, n = self.strategy_dict[idx]  # input, output pruning ratio
+            flops += self.layer_info_dict[idx]['flops'] * c * n
+            # add buffer computation
+            flops += self._get_buffer_flops(idx) * c  # only related to input channel reduction
+        return flops
+
+    def _cur_reduced(self):
+        # return the reduced weight
+        reduced = self.org_flops - self._cur_flops()
+        return reduced
+
+    def _build_index(self):
+        self.prunable_idx = []
+        self.prunable_ops = []
+        self.layer_type_dict = {}
+        self.strategy_dict = {}
+        self.buffer_dict = {}
+        this_buffer_list = []
+        self.org_channels = []
+        # build index and the min strategy dict
+        for i, m in enumerate(self.model.modules()):
+            if isinstance(m, PrunerModuleWrapper):
+                m = m.module
+                if type(m) == nn.Conv2d and m.groups == m.in_channels:  # depth-wise conv, buffer
+                    this_buffer_list.append(i)
+                else:  # really prunable
+                    self.prunable_idx.append(i)
+                    self.prunable_ops.append(m)
+                    self.layer_type_dict[i] = type(m)
+                    self.buffer_dict[i] = this_buffer_list
+                    this_buffer_list = []  # empty
+                    self.org_channels.append(m.in_channels if type(m) == nn.Conv2d else m.in_features)
+
+                    self.strategy_dict[i] = [self.lbound, self.lbound]
+
+        self.strategy_dict[self.prunable_idx[0]][0] = 1  # modify the input
+        self.strategy_dict[self.prunable_idx[-1]][1] = 1  # modify the output
+
+        self.shared_idx = []
+        if self.args.model_type == 'mobilenetv2':  # TODO: to be tested! Share index for residual connection
+            connected_idx = [4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32]  # to be partitioned
```

Comment:
what are these magic numbers?

Reply:
This is hard coded mobilenetv2 layers with residual connections, can be reimplemented later using our shape dependency code.